client/src/components/FilesDialog/FilesDialog.test.ts

```diff
@@ -285,124 +239,156 @@ describe("FilesDialog, directory mode", () => {
+        const backBtn = this.getBackButton();
+        await backBtn.trigger("click");
         await flushPromises();
     }
 
-    assertAllRenderedItemsSelected() {
-        this.wrapper.vm.items.forEach((item: RowElement) => {
-            // assert style/icon of selected field
-            expect(item._rowVariant).toBe(selectionStates.selected);
-            // selected items are selected
-            expect(
-                item.isLeaf
-                    ? this.wrapper.vm.model.exists(item.id)
-                    : this.wrapper.vm.selectedDirectories.some(({ id }: RowElement) => id === item.id)
-            ).toBe(true);
-        });
+    async openDirectoryById(directoryId: string) {
+        const directory = this.findRenderedDirectory(directoryId);
+        return this.openDirectory(directory);
     }
 
-    async openDirectory(directoryId: string) {
-        this.getTable().$emit("open", this.getRenderedDirectory(directoryId));
+    async openDirectory(directory: RowElement) {
+        this.getTable().vm.$emit("open", directory);
         await flushPromises();
     }
 
     async clickOn(element: Element) {
-        this.getTable().$emit("clicked", element);
+        this.getTable().vm.$emit("clicked", element);
+        await flushPromises();
+    }
+
+    async selectAll() {
+        this.getTable().vm.$emit("toggleSelectAll");
         await flushPromises();
     }
 
-    getRenderedDirectory(directoryId: string): RowElement {
-        return this.wrapper.vm.items.find(({ id }: RowElement) => directoryId === id);
+    findRenderedDirectory(directoryId: string): RowElement {
+        const directory = this.getRenderedRows().find(({ id }: RowElement) => directoryId === id);
+        if (!directory) {
+            throw new Error(`Directory with id ${directoryId} not found`);
+        }
+        return directory;
     }
 
     getRenderedFiles(): RowElement[] {
-        return this.wrapper.vm.items.filter((item: RowElement) => item.isLeaf);
+        return this.getRenderedRows().filter((item: RowElement) => item.isLeaf);
+    }
+
+    findFirstFile(): RowElement {
+        const file = this.getRenderedFiles().at(0);
+        if (!file) {
+            throw new Error("File not found");
+        }
+        return file;
+    }
+
+    async applyToEachFile(func: (item: RowElement) => void) {
+        this.getRenderedFiles().forEach((item) => {
+            func(item);
+        });
+        await flushPromises();
+    }
+
+    getTable(): any {
+        return this.wrapper.findComponent(DataDialogTable);
+    }
+
+    getButtonById(id: string): any {
+        const button = this.wrapper.find(id);
+        expect(button.exists()).toBe(true);
+        return button;
+    }
+
+    getOkButton(): any {
+        return this.getButtonById("#ok-btn");
+    }
+
+    getBackButton(): any {
+        return this.getButtonById("#back-btn");
+    }
+
+    getRenderedRows(): RowElement[] {
+        return this.getTable().vm.items as RowElement[];
```

Comment:
@ElectronicBlueberry, in relation to https://github.com/galaxyproject/galaxy/pull/16391#discussion_r1267422695, I have rewritten these tests to completely avoid exposing the `FilesDialog` component internals while still using `shallowMount`.

But I'm using this hacky line here to access the internal `items` of the stubbed child component to be able to test the state of the items, so I feel dirty again :laughing: Do you have any idea of how to avoid this? maybe creating a custom stub that renders the items in the HTML? or would it be overkill? :thinking:

Reply:
Since this is a stub, this seems fine to me. Another option may be to use `.props` on the stub, which would make it clearer what you're accessing, but I'm not sure if that works on stubs.